File: netlify/functions/chatbot.js
import { OpenAI } from "openai";
import { tavily } from "@tavily/core";

// Inline filter function
function isQuestionValid(userInput) {
  const bannedKeywords = [
    "joke", "funny", "lol", "haha", "laugh",
    "crush", "kiss", "hug", "flirt",
    "dating", "sex",
    "boyfriend", "girlfriend",
    "do you love me", "tell me a joke", "marry me", "i love you",
    "chat with me", "best friend", "are you single", "romantic", "cute"
  ];

  const userInputLower = userInput.toLowerCase();
  for (const word of bannedKeywords) {
    if (userInputLower.includes(word)) {
      return false;
    }
  }
  return true;
}

// Remove citations like   or [4:0†source]
function removeCitations(text) {
  return text
    .replace(/【\d+:\d+†[^】]+】/g, "") // match any characters until the closing 】
    .replace(/\[\d+:\d+†[^\]]+\]/g, ""); // match any characters until the closing ]
}

export async function handler(event, context) {
  if (event.httpMethod !== "POST") {
    return {
      statusCode: 405,
      body: JSON.stringify({ error: "Method Not Allowed" }),
      headers: { "Content-Type": "application/json" },
    };
  }

  try {
    const body = JSON.parse(event.body || "{}");
    const query = (body.query || "").trim();
    let thread_id = body.thread_id;

    if (!query || !isQuestionValid(query)) {
      return {
        statusCode: 200,
        body: JSON.stringify({
          thread_id,
          response:
            "This question is not appropriate or relevant. Please ask something based on your role or documents.",
        }),
        headers: { "Content-Type": "application/json" },
      };
    }

    const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
    const TAVILY_API_KEY = process.env.TAVILY_API_KEY;
    const VECTOR_STORE_ID = process.env.VECTOR_STORE_ID;
    const ASSISTANT_ID = process.env.ASSISTANT_ID;

    if (!OPENAI_API_KEY || !TAVILY_API_KEY || !VECTOR_STORE_ID || !ASSISTANT_ID) {
      return {
        statusCode: 500,
        body: JSON.stringify({ error: "Missing required environment variables" }),
        headers: { "Content-Type": "application/json" },
      };
    }

    const client = new OpenAI({ apiKey: OPENAI_API_KEY });
    const tavilyClient = tavily({ apiKey: TAVILY_API_KEY });

    async function tavilySearch(q) {
      try {
        const response = await tavilyClient.search(q);

        if (response.answer) return response.answer;

        const results = response.results || [];
        if (results.length > 0) {
          return results.map((r) => `${r.title}: ${r.url}`).join("\n");
        }

        return "No relevant results found on the web.";
      } catch (error) {
        return `Error performing Tavily search: ${error.message}`;
      }
    }

    if (!thread_id) {
      const thread = await client.beta.threads.create();
      thread_id = thread.id;
    }

    await client.beta.threads.messages.create(thread_id, {
      role: "user",
      content: query,
    });

    let run = await client.beta.threads.runs.createAndPoll(thread_id, {
      assistant_id: ASSISTANT_ID,
    });

    if (run.required_action?.type === "submit_tool_outputs") {
      const tool_outputs = await Promise.all(
        run.required_action.submit_tool_outputs.tool_calls.map(async (action) => {
          const functionName = action.function.name;
          const args = JSON.parse(action.function.arguments);

          let result;
          if (functionName === "tavily_search") {
            result = await tavilySearch(args.query || "");
          } else {
            result = "Unknown tool requested.";
          }

          return {
            tool_call_id: action.id,
            output: result,
          };
        })
      );

      run = await client.beta.threads.runs.submitToolOutputsAndPoll(
        thread_id,
        run.id,
        { tool_outputs }
      );
    }

    const messagesResponse = await client.beta.threads.messages.list(thread_id);

    const assistantMessages = messagesResponse.data.filter(
      (m) => m.role === "assistant"
    );

    if (assistantMessages.length === 0) {
      return {
        statusCode: 500,
        body: JSON.stringify({ error: "No response from assistant" }),
        headers: { "Content-Type": "application/json" },
      };
    }

    let assistantResponse = assistantMessages[0].content[0].text.value;

    // Remove citation markers
    assistantResponse = removeCitations(assistantResponse);

    return {
      statusCode: 200,
      body: JSON.stringify({ thread_id, response: assistantResponse }),
      headers: { "Content-Type": "application/json" },
    };
  } catch (error) {
    return {
      statusCode: 500,
      body: JSON.stringify({ error: error.message }),
      headers: { "Content-Type": "application/json" },
    };
  }
}

File: netlify/functions/chatbot.py
import json
import os
import requests

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
TAVILY_API_KEY = os.getenv("TAVILY_API_KEY")
ASSISTANT_ID = os.getenv("ASSISTANT_ID")
VECTOR_STORE_ID = os.getenv("VECTOR_STORE_ID")

BLOCKED_WORDS = ["hack", "illegal", "kill"]

OPENAI_BASE_URL = "https://api.openai.com/v1"

HEADERS_OPENAI = {
    "Authorization": f"Bearer {OPENAI_API_KEY}",
    "Content-Type": "application/json"
}

HEADERS_TAVILY = {
    "Authorization": f"Bearer {TAVILY_API_KEY}",
    "Content-Type": "application/json"
}

def is_question_valid(query):
    q = query.lower()
    return not any(word in q for word in BLOCKED_WORDS)

def tavily_search(query):
    try:
        payload = {
            "query": query,
            "search_depth": "basic",
            "include_answer": True,
            "include_raw_content": False
        }
        response = requests.post("https://api.tavily.com/search", headers=HEADERS_TAVILY, json=payload)
        response.raise_for_status()
        data = response.json()
        return data.get("answer", "No relevant results found.")
    except Exception as e:
        print("Tavily search error:", e)
        return "Error fetching search results."

def normalize_thread_id(raw):
    if raw is None:
        return None
    if isinstance(raw, str):
        return raw.strip()
    if isinstance(raw, dict):
        if "id" in raw and isinstance(raw["id"], str):
            return raw["id"].strip()
        first_val = next(iter(raw.values()), None)
        if isinstance(first_val, str):
            return first_val.strip()
    return str(raw).strip()

def handler(event, context):
    if event["httpMethod"] != "POST":
        return {
            "statusCode": 405,
            "body": json.dumps({"error": "Method not allowed"})
        }
    
    try:
        content_type = event["headers"].get("content-type") or event["headers"].get("Content-Type") or ""
        if "application/json" in content_type:
            body = json.loads(event.get("body") or "{}")
        else:
            # handle form-url-encoded
            from urllib.parse import parse_qs
            body = parse_qs(event.get("body") or "")
            # parse_qs returns lists per key
            body = {k: v[0] for k,v in body.items()}

        query = body.get("query", "").strip()
        thread_id = normalize_thread_id(body.get("thread_id"))

        # Validate thread_id is string or None
        if thread_id and not isinstance(thread_id, str):
            thread_id = str(thread_id)
        thread_id = thread_id.strip() if thread_id else None

        print(f"Received query: {query}")
        print(f"Thread ID: {thread_id}")

        if not query or not is_question_valid(query):
            return {
                "statusCode": 200,
                "body": json.dumps({
                    "thread_id": thread_id or None,
                    "response": "This question is not appropriate or relevant. Please ask something based on your role or documents."
                })
            }

        if not ASSISTANT_ID or not VECTOR_STORE_ID:
            return {
                "statusCode": 500,
                "body": json.dumps({"error": "Missing VECTOR_STORE_ID or ASSISTANT_ID"})
            }

        # Create a new thread if no thread_id provided
        if not thread_id:
            url = f"{OPENAI_BASE_URL}/assistants/{ASSISTANT_ID}/threads"
            resp = requests.post(url, headers=HEADERS_OPENAI)
            resp.raise_for_status()
            data = resp.json()
            thread_id = data.get("id")
            if not thread_id:
                raise Exception("Failed to create a new thread")
            print(f"Created new thread: {thread_id}")

        # Send user message
        url_send_msg = f"{OPENAI_BASE_URL}/threads/{thread_id}/messages"
        payload_msg = {
            "role": "user",
            "content": query
        }
        resp = requests.post(url_send_msg, headers=HEADERS_OPENAI, json=payload_msg)
        resp.raise_for_status()

        # Run assistant
        url_run = f"{OPENAI_BASE_URL}/assistants/{ASSISTANT_ID}/threads/{thread_id}/runs"
        resp = requests.post(url_run, headers=HEADERS_OPENAI)
        resp.raise_for_status()
        run_data = resp.json()

        # Handle tool calls if required
        if run_data.get("required_action", {}).get("type") == "submit_tool_outputs":
            tool_outputs = []
            for call in run_data["required_action"]["submit_tool_outputs"]["tool_calls"]:
                fn_name = call["function"]["name"]
                args = json.loads(call["function"]["arguments"])
                output = "Unknown tool"

                if fn_name == "tavily_search":
                    output = tavily_search(args.get("query", ""))

                tool_outputs.append({
                    "tool_call_id": call["id"],
                    "output": output
                })

            # Submit tool outputs and poll again
            url_submit_tools = f"{OPENAI_BASE_URL}/threads/{thread_id}/runs/{run_data['id']}/tool_outputs"
            resp = requests.post(url_submit_tools, headers=HEADERS_OPENAI, json={"tool_outputs": tool_outputs})
            resp.raise_for_status()
            run_data = resp.json()

        # Get messages for thread
        url_messages = f"{OPENAI_BASE_URL}/threads/{thread_id}/messages"
        resp = requests.get(url_messages, headers=HEADERS_OPENAI)
        resp.raise_for_status()
        messages_data = resp.json()

        # Extract assistant responses
        assistant_messages = [m for m in messages_data.get("data", []) if m.get("role") == "assistant"]
        final_response = "\n".join(
            m.get("content", {}).get("text", {}).get("value", "") for m in assistant_messages
        ).strip()

        if not final_response:
            final_response = "⚠️ Assistant did not return a message."

        return {
            "statusCode": 200,
            "body": json.dumps({
                "thread_id": thread_id,
                "response": final_response
            })
        }

    except Exception as e:
        print("Error in chatbot function:", e)
        return {
            "statusCode": 500,
            "body": json.dumps({"error": str(e)})
        }
